### ISSUE TITLE
IFCELEMENT attribute definitions 

### ISSUE THREAD

**Issue by Hirak2019** [Autodesk/revit-ifc#145]
Hello, 
I have a remark 

For the ifc export the attributes of the ifcelement are defined as follows:
IFCELEMENT('GUID',OwnerHistory,'**Name:ObjectType:Tag**',Description,'**Name:ObjectType**',#ObjectPlacement,#Representation,'Tag',.ElementType.);

While in the standard IFC4 we have that
IFCELEMENT('GUID',OwnerHistory,'**Name**',Description,'**ObjectType**',#ObjectPlacement,#Representation,'Tag',.ElementType.);

Thank you.

**Comment by AngelVelezSosa** [comment]
Most Revit elements don't have names.  As such, we create their name as you state above.  You can give an element a name if you use IfcName (for example).  If you used just "Name" as stated above, all walls would be called "Walls", which is likely what no one wants.

**Comment by Hirak2019** [comment]
for wall for example the familly name is exported as **Name** and the type as **ObjectType** 
![image](https://user-images.githubusercontent.com/58464538/70056518-e27bb280-15db-11ea-855d-3fb68c5e07c2.png)

**Comment by AngelVelezSosa** [comment]
That's correct and the default behavior unless you override it.

**Comment by ArjanPeeters** [comment]
Could there be a way changing the name centrally instead of for every item? or maybe per category?

**Comment by AngelVelezSosa** [comment]
OK, say you have 100 walls in the project of the same type (for simplicity, System Family: Basic Wall, Generic - 200mm as shown above).  What would you want their names to be?

**Comment by Hirak2019** [comment]
Name will be **"Basic Wall"** and objectType will be **"Generic - 200mm"** for the 100 walls. 
the only difference it will be the **tag** 

there is no problem that the walls have the same name and the same type on the contrary all this is very possible

**Comment by ArjanPeeters** [comment]
I would like to be able to make my own naming convention. In the Netherlands we use the nlsfb classification. Most names start with something like 22_drywall_100mm ore something. So naming convention like [first 2 chars of nlsfb]_[type]_[with parameter] would be very helpfull

**Comment by AngelVelezSosa** [comment]
With the IfcName parameter and Dynamo (or a Revit macro), I think it would be  fairly easy to come up with any naming convention you wanted  You would just run your macro before any export to update names (for anything unnamed, presumably) and then export.

**Comment by ArjanPeeters** [comment]
Have you tried rolling out a dynamo script for a whole office. Or a country. I don’t think it’s a valid option.

**Comment by AngelVelezSosa** [comment]
Don't you have to roll out something to have a customizable naming convention?  How would you ideally do this?

**Comment by tsolsen** [comment]
Being able to map OOTB parameters to Root parameters would be a great help, and cut down considerably on our need for IFC parameters.

I would do something like:
```
Type Name = IfcObjectType
Type Mark = IfcName (Type)
Type Comments = IfcDescription (Type)
Mark = IfcName
Comments = IfcDescription
```

**Comment by ArjanPeeters** [comment]
Maybe even being able to do something like
IfcTypeName = {assemby code[:2] +’_’+Type Name}.

Or being. Able to use python code for naming , ;-)

**Comment by AngelVelezSosa** [comment]
Something like that may already work.  @WawanSolihin ?

**Comment by dvrvb** [comment]
A lot of stuff is indeed already possible …
<img width="961" alt="image" src="https://user-images.githubusercontent.com/41319048/70096321-1c25db00-1627-11ea-9d45-919e8bf3742b.png">

**Comment by ArjanPeeters** [comment]
Is this also possible for system categories?

**Comment by dvrvb** [comment]
yes, indeed
same sample, but now the (basic) wall is selected
<img width="554" alt="image" src="https://user-images.githubusercontent.com/41319048/70117197-1949db00-1665-11ea-9467-dd031d0d1df9.png">

**Comment by ArjanPeeters** [comment]
Is this in somewhere in the manual?

**Comment by dvrvb** [comment]
yes, in the one that I'm preparing to publish 'soon' (but unfortunately not anymore before XMAS)

**Comment by ArjanPeeters** [comment]
Now I’m excited. Where do you publish this stuff? Here on github?

**Comment by dvrvb** [comment]
Nope, it will be published on a 'regular' site. I will let you know when it is going 'on air'.

**Comment by Hirak2019** [comment]
another remark
in the ifc4 exported from revit the IFCSPACE is exported like 
IFCSPACE('GUID',#42,'**Number**',$,$,ObjectPlacement,Representation,'**LongName**',.ELEMENT.,.SPACE.,$);
![image](https://user-images.githubusercontent.com/58464538/70128160-b5cba780-167c-11ea-946a-1cf929d34428.png)

it should be exported like this (reference to the IFC4 Standard https://standards.buildingsmart.org/IFC/RELEASE/IFC4_1/FINAL/HTML/schema/ifcproductextension/lexical/ifcspatialelement.htm)
IFCSPACE('GUID',#42,'**Name**',$,$,ObjectPlacement,Representation,'**LongName**',.ELEMENT.,.SPACE.,$);

**Comment by dvrvb** [comment]
Maybe it can be confusing, but it is the correct mapping:
Revit/Solibri Number = IFC Definition Attribute Name
Revit/Solibri Name = IFC Definition Attribute LongName

**Comment by WawanSolihin** [comment]
Regarding the IfcSpace, the schema documentation says it very clearly:
_The following guidelines should apply for using the Name, Description, LongName and ObjectType attributes.
Name holds the unique name (or space number) from the plan.
Description holds any additional information field the user may have specified, there are no further recommendations.
LongName holds the full name of the space, it is often used in addition to the Name, if a number is assigned to the room, then the descriptive name is exchanged as LongName.
ObjectType holds the space type, i.e. usually the functional category of the space ._

**Comment by WawanSolihin** [comment]
> A lot of stuff is indeed already possible …

Dirk has said it all. The feature to define a simple expression to combine values from multiple other parameters and/or a simple formula can be done. The basic documentation is available here: https://sourceforge.net/p/ifcexporter/wiki/Notes%20on%20parameter%20expression/

**Comment by tsolsen** [comment]
> The feature to define a simple expression to combine values from multiple other parameters and/or a simple formula can be done. The basic documentation is available here: https://sourceforge.net/p/ifcexporter/wiki/Notes%20on%20parameter%20expression/

As I alluded to in my comment, the feature - while there - is not widely known. How could that be improved?
I tried today to enter the formula into the formula field, to lock it down - this was not allowed.

The feature is super nice, and could be used all over the software. But it is only usable on export, not in schedules and annotations.
I'm sure many users would like built in concatenate, even if it was written in C#(?) code.

Finally - perhaps the documentation could be moved to a wiki or markdown file, hosted on GitHub?
That way we could make pull request against it, and perhaps give a hand in improving.
See also #50
